Seeing "Promise rejected with error: undefined" for every failed `browser.test.assert...` call

kiaraarose opened the pull request:
#### b9c30151741e228bbc11dc2034e9f6f5f4a9579d
<pre>
Seeing &quot;Promise rejected with error: undefined&quot; for every failed `browser.test.assert...` call
<a href="https://bugs.webkit.org/show_bug.cgi?id=292641">https://bugs.webkit.org/show_bug.cgi?id=292641</a>
<a href="https://rdar.apple.com/150804691">rdar://150804691</a>

Reviewed by Timothy Hatcher.

Use the message passed along to the browser.test.assert call if an error isn&apos;t generated when the
assertion failed. If a message isn&apos;t passed along, then return &quot;Promise rejected without an error&quot;.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
(WebKit::WebExtensionAPITest::assertTrue):
(WebKit::WebExtensionAPITest::assertDeepEq):
(WebKit::WebExtensionAPITest::assertEquals):
(WebKit::WebExtensionAPITest::startNextTest):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h:
(WebKit::WebExtensionAPITest::recordAssertionIfNeeded):

Canonical link: <a href="https://commits.webkit.org/294629@main">https://commits.webkit.org/294629@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/856206cbc9fd842500c770d19b206cfcb9f5049a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102517 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/22186 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12500 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107678 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/53154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22494 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30692 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77988 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/53154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105523 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/17426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/92534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58324 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/17263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52511 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/10630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110053 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/29649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/21865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/30013 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/88727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/86565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/9117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/23899 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16652 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/29577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/34889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/29388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/32711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/30947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -438,8 +438,17 @@
+            else
+                errorMessage = @"Promise rejected without an error";
+
+            WebProcess::singleton().send(Messages::WebExtensionController::TestFinished(test.testName, false, errorMessage, test.location.first, test.location.second), test.webExtensionControllerIdentifier);
```
i'm not seeing this extra newline locally, which is strange

xeenon replied:
I think the newline is fine.

elijahsawyers commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h:
```diff
@@ -88,14 +88,17 @@ class WebExtensionAPITest : public WebExtensionAPIObject, public JSWebExtensionW
+    void recordAssertionIfNeeded(bool result, const String& message)
     {
-        if (m_runningTest && !m_hitAssertion && !result)
+        if (m_runningTest && !m_hitAssertion && !result) {
```
Can we early return to avoid the nested condition?

If we're not running a test, we're running a test and already hit an assertion, or we're running a test and the assertion passed, we don't need to update the assertion.

I don't feel strongly about this. It's arguably less readable.

```Objective-C
if (!m_runningTest || (m_runningTest && m_hitAssertion) || (m_runningTest && result))
    return;

m_hitAssertion = true;
m_assertionMessage = message;
```

kiaraarose replied:
since we plan on aborting the function if we already hit an assertion then `(m_runningTest && m_hitAssertion)` won't apply. i think this makes the most sense
```
if (!m_runningTest || (m_runningTest && result))
    return;
    
m_hitAssertion = true;
m_assertionMessage = message;
```

kiaraarose replied:
or 
```
if (m_runningTest && !result) {
    m_hitAssertion = true;
    m_assertionMessage = message;
}
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -438,8 +438,17 @@
+            } else if (!m_assertionMessage.isNull())
+                errorMessage = adoptNS([[NSString alloc] initWithFormat:@"Promise rejected with an error: %@", m_assertionMessage.createNSString().get()]).get();
+            else
+                errorMessage = @"Promise rejected without an error";
```
I'm wondering if we should just set the message as-is and not prefix with "Promise rejected with an error:". Seems odd to bake that in here.

kiaraarose replied:
i kind of like the prefix. at least in terms of readability of the error in WPT. this is what it'd look like
`safari-web-extension://D39F4C09-2D13-406E-AAB8-75C89823321C/background.js:10. Promise rejected with an error: expected different tab id`

elijahsawyers commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
```diff
@@ -438,8 +438,17 @@
-            JSValue *errorMessageValue = error.isObject && [error hasProperty:@"message"] ? error[@"message"] : error;
-            WebProcess::singleton().send(Messages::WebExtensionController::TestFinished(test.testName, false, adoptNS([[NSString alloc] initWithFormat:@"Promise rejected with an error: %@", debugString(errorMessageValue)]).get(), test.location.first, test.location.second), test.webExtensionControllerIdentifier);
+            NSString *errorMessage;
+            if (error) {
```
Can this logic be cleaned up by using `combineMessages`?

kiaraarose committed:
Seeing "Promise rejected with error: undefined" for every failed `browser.test.assert...` call
https://bugs.webkit.org/show_bug.cgi?id=292641
rdar://150804691

Reviewed by Timothy Hatcher.

Use the message passed along to the browser.test.assert call if an error isn't generated when the
assertion failed. If a message isn't passed along, then return "Promise rejected without an error".

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
(WebKit::WebExtensionAPITest::assertTrue):
(WebKit::WebExtensionAPITest::assertDeepEq):
(WebKit::WebExtensionAPITest::assertEquals):
(WebKit::WebExtensionAPITest::startNextTest):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h:
(WebKit::WebExtensionAPITest::recordAssertionIfNeeded):

Canonical link: https://commits.webkit.org/294629@main
